sale_order_type/models/sale.py

```diff
@@ -74,6 +75,19 @@ def _prepare_invoice(self):
             res["sale_type_id"] = self.type_id.id
         return res
 
+    @api.constrains("company_id")
+    def _check_so_type_company(self):
+        if self.filtered(
+            lambda r: r.type_id.company_id
+            and r.company_id
+            and r.type_id.company_id != r.company_id
+        ):
+            raise ValidationError(_("Document's company and type's company mismatch"))
+
+    @api.onchange("company_id")
+    def _onchange_company(self):
+        self.type_id = False
+
```

Comment:
```suggestion
```

`_compute_sale_type_id` already depends on `company_id`, and it's supposed to do a better job (setting the sale_type of the partner)

Why would we need this onchange?


---

I understand that `_compute_sale_type_id` is a bit flawed.
Meaning, take this scenario for instance:


Sale Type Company A has company_id = "Company A"
Partner A has sale_type = "Sale Type Company A"
We are logged in in "Company B", create a sale.order and select partner "Partner A".

The sale order's type will be "Sale Type Company A", but it would break check_company rules.


IMO we should transform the sale_type_id field in res.partner a to a property_sale_type_id (company_dependant).